File: python/nart/core/art/Proxy.py
# Copyright 2022 SenseTime Group Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from ... import _nart
from . import FakeOp
import numpy as np

import logging

logger = logging.getLogger("nart.core.art")

cuda_workspaces = None
default_workspace = None
quant_workspace = None


# class DeserializeConfig(_nart.art.DeserializeConfig):
#     def __init__(self, workspaces=None, input_mem_tp=None, backend='default') :
#         super(DeserializeConfig, self).__init__()
#         if not workspaces:
#             global default_workspace, quant_workspace
#             default_workspace = _nart.art.Workspace.new_default()
#             quant_workspace = _nart.art.Workspace.new_quant()
#             workspaces = [default_workspace, quant_workspace]
#         if not input_mem_tp:
#             input_mem_tp = default_workspace.mem_tp
#         cuda_enabled = hasattr(_nart.art.Workspace, "new_cuda")
#         if not cuda_enabled and backend == "cuda":
#             logger.warning("Requesting cuda workspace, but cuda is not built, fallback to default only")
#             backend = "default"

#         if backend == 'cuda':
#             global cuda_workspace
#             cuda_workspace = _nart.art.Workspace.new_cuda()
#             cuda_mem_tp = cuda_workspace.mem_tp
#             self.workspaces = [cuda_workspace] + list(workspaces)
#             self.input_mem_tp = cuda_mem_tp
#         else:
#             self.workspaces = list(workspaces)
#             self.input_mem_tp = input_mem_tp


class DeserializeConfig:
    def __init__(self) -> None:
        raise NotImplementedError("")


class ModuleProxy:
    class Constraint:
        __slot__ = ["name", "item", "dtype", "constraint", "default_value"]

        def __init__(self, name, item, dtype, constraint, default_value=None):
            self.name = name
            self.item = item
            self.dtype = dtype.lower()
            self.constraint = constraint
            if constraint == "optional":
                assert default_value is not None
                self.default_value = default_value

        def __repr__(self):
            if self.constraint == "optional":
                return f'{{name="{self.name}", dtype="{self.dtype}", constraint="{self.constraint}", default_value="{self.default_value}"}}'
            return f'{{name="{self.name}", dtype="{self.dtype}", constraint="{self.constraint}"}}'

    def __init__(self, module_name):
        self.true_obj = _nart.proxy("default")
        self.mp_name_tp_ = {}
        self.mp_name_constraint_ = {}
        for item in self.true_obj.op_tps:
            self.mp_name_tp_[item[1]] = item[0]
            self.mp_name_constraint_[item[1]] = {}
            for itm in self.true_obj.constraints_of(item[1]):
                self.mp_name_constraint_[item[1]][itm[0]] = ModuleProxy.Constraint(*itm)
        self.repeated_setting = def_setting_cls("repeate")
        self.single_setting = def_setting_cls("single")

        for item in self.mp_name_tp_:
            setattr(self, item, gen_class(self, item))

    @property
    def op_tps(self):
        return list(self.mp_name_tp_.keys())

    def constraints_of(self, key):
        return self.mp_name_constraint_[key] if key in self.mp_name_constraint_ else []


def def_setting_cls(rep):
    class setting:
        def __init__(self, name, dtype, v):
            if rep is True:
                self.data = np.array(v, dtype=dtype)
            else:
                if hasattr(v, "__len__"):
                    assert 1 == len(v)
                else:
                    v = [v]
                self.data = np.array(v, dtype=dtype)
            self.dtype = dtype

    setting.rep = rep
    return setting


def gen_class(mod, name):
    class _cls(FakeOp.FakeOpInterface):
        def __init__(self, inputs, params=None, layer_name=None, **k):
            super(_cls, self).__init__(inputs, params)
            self.settings = {}
            self.layer_name = layer_name
            for item in mod.mp_name_constraint_[name].values():
                attr_name = item.name.lower()[8:]
                if item.constraint == "required" and attr_name not in k:
                    err_str = f"{name} should set attr(s) "
                    err_str += str(
                        list(
                            map(
                                lambda x: x.name.lower()[8:],
                                filter(
                                    lambda x: x.constraint == "required",
                                    mod.mp_name_constraint_[name].values(),
                                ),
                            )
                        )
                    )
                    raise RuntimeError(err_str)
            for key, v in k.items():
                setattr(self, key, v)
            _nart.proxy.proxy_init(self)

        def op_tp_code(self):
            return mod.mp_name_tp_[name]

        def get_settings(self):
            res = []
            for k, v in self.settings.items():
                res.append(
                    (mod.mp_name_constraint_[name]["SETTING_" + k.upper()].item, v)
                )
            return res

        def __repr__(self):
            res = "FakeOp({self.__class__.__name__}):\n"
            for k, v in self.settings.items():
                res += f"\tsetting {k} -> {v}\n"
            res += "\tpass"
            return res

        @property
        def module(self):
            return mod

    def gen_attr_fun(attr_name, item):
        def attr_getter(self):
            if attr_name in self.settings:
                return self.settings[attr_name]
            if item.constraint == "optional":
                return item.default_value

        def attr_setter(self, value):
            if item.constraint != "repeated":
                self.settings[attr_name] = np.array([value], dtype=item.dtype)[0]
            else:
                self.settings[attr_name] = np.array(value, dtype=item.dtype)

        return attr_getter, attr_setter

    for item in mod.mp_name_constraint_[name].values():
        attr_name = item.name.lower()[8:]
        setattr(_cls, attr_name, property(*gen_attr_fun(attr_name, item)))
    _cls.__name__ = name
    return _cls

File: python/src/nart/proxy.cpp
/*
 * Copyright 2022 SenseTime Group Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "art/op_settings.h"

#include "include/fake.h"
#include "include/fakes.h"

#include "proxy.h"

#include "./python_fakes.h"
#include "./symbol.h"

namespace py = pybind11;
namespace nart {
namespace art {

    static py::object data_tr_optional(uint32_t dtype, uvalue_t v)
    {
        switch (dtype) {
        case dtUINT8:
            return py::cast(v.u8);
        case dtUINT16:
            return py::cast(v.u16);
        case dtUINT32:
            return py::cast(v.u32);
        case dtUINT64:
            return py::cast(v.u64);
        case dtINT8:
            return py::cast(v.i8);
        case dtINT16:
            return py::cast(v.i16);
        case dtINT32:
            return py::cast(v.i32);
        case dtINT64:
            return py::cast(v.i64);

        case dtFLOAT32:
            return py::cast(v.f32);

        case dtBOOL:
            return py::cast(v.b);

        case dtSTR:
            return py::cast(v.str);

        case dtCPTR:
            return py::cast(v.cptr);
        }
        throw std::runtime_error(
            std::string("cannot cast datatype ") + datatype_name_from_type(dtype));
    }

    /*
     * proxy of `module_t` in art, through which a list of supported op could be automaticly
     * constructed
     */
    class python_module_proxy {
    public:
        python_module_proxy(const module_t *module) : module_(module) { }
        uint32_t op_group() const { return module_->op_group; }
        op_tp_entry_t *search_op_tp_entry(std::string op_tp) const
        {
            for (op_tp_entry_t *p = module_->op_tp_entry; p && p->tp; ++p) {
                if (op_tp == p->tp->name)
                    return p;
            }
            return nullptr;
        }
        py::list op_tps() const
        {
            py::list res;
            for (op_tp_entry_t *p = module_->op_tp_entry; p && p->tp; ++p) {
                res.append(py::make_tuple(p->tp->op_tp_code, std::string(p->tp->name)));
            }
            return res;
        }
        py::list op_tp_constraints(std::string op_tp) const
        {
            auto p = search_op_tp_entry(op_tp);
            if (p->tp->name == op_tp) {
                /* list<(name:string, item:int, datatype:string, constraint:string [,
                 * default_value:object])>
                 */
                py::list res;
                for (const setting_constraint_t *pc = p->tp->constraints; pc->item != SETTING_END;
                     pc++) {
                    if (pc->ctp == ENUM_SETTING_CONSTRAINT_OPTIONAL) {
                        res.append(py::make_tuple(
                            pc->name, pc->item, datatype_name_from_type(pc->dtype), "optional",
                            data_tr_optional(pc->dtype, pc->constraint.optional.default_value)));
                    } else {
                        res.append(py::make_tuple(
                            pc->name, pc->item, datatype_name_from_type(pc->dtype),
                            [](enum_setting_constraint_tp_t ctp) {
                                switch (ctp) {
                                case ENUM_SETTING_CONSTRAINT_REQUIRED:
                                    return "required";
                                case ENUM_SETTING_CONSTRAINT_OPTIONAL:
                                    return "optional";
                                case ENUM_SETTING_CONSTRAINT_REPEATED:
                                    return "repeated";
                                default:
                                    return "error";
                                }
                            }(pc->ctp)));
                    }
                }
                return res;
            }
            throw std::runtime_error("cannot find op type '" + op_tp + "'");
        }

    private:
        const module_t *module_;
    };

    static std::unique_ptr<fake_op::fake_setting>
    parse_setting(const setting_constraint_t *constraint, py::handle obj)
    {
        using res_tp = std::unique_ptr<fake_op::fake_setting>;
        if (constraint->ctp == ENUM_SETTING_CONSTRAINT_REPEATED) {
            /* repeat */
            switch (constraint->dtype) {
            case dtUINT8:
                return res_tp(
                    new fake_op::fake_setting_repeat_tp<uint8_t>(obj.cast<vector<uint8_t>>()));
            // case dtUINT16:
            //     return res_tp(new
            //     fake_op::fake_setting_repeat_tp<uint16_t>(obj.cast<vector<uint16_t>>()));
            case dtUINT32:
                return res_tp(
                    new fake_op::fake_setting_repeat_tp<uint32_t>(obj.cast<vector<uint32_t>>()));
            // case dtINT8:
            //     return res_tp(new
            //     fake_op::fake_setting_repeat_tp<int8_t>(obj.cast<vector<int8_t>>()));
            // case dtINT16:
            //     return res_tp(new
            //     fake_op::fake_setting_repeat_tp<int16_t>(obj.cast<vector<int16_t>>()));
            case dtINT32:
                return res_tp(
                    new fake_op::fake_setting_repeat_tp<int32_t>(obj.cast<vector<int32_t>>()));
            case dtFLOAT32:
                return res_tp(
                    new fake_op::fake_setting_repeat_tp<float>(obj.cast<vector<float>>()));
            case dtSTR:
                return res_tp(
                    new fake_op::fake_setting_repeat_tp<string>(obj.cast<vector<string>>()));
                // case dtFLOAT64:
                //     return res_tp(new
                //     fake_op::fake_setting_repeat_tp<double>(obj.cast<vector<double>>()));
            }
        } else {
            /* single */
            switch (constraint->dtype) {
            case dtBOOL:
                return res_tp(new fake_op::fake_setting_tp<bool>(obj.cast<bool>()));
            case dtUINT8:
                return res_tp(new fake_op::fake_setting_tp<uint8_t>(obj.cast<uint8_t>()));
            case dtUINT16:
                return res_tp(new fake_op::fake_setting_tp<uint16_t>(obj.cast<uint16_t>()));
            case dtUINT32:
                return res_tp(new fake_op::fake_setting_tp<uint32_t>(obj.cast<uint32_t>()));
            case dtINT8:
                return res_tp(new fake_op::fake_setting_tp<int8_t>(obj.cast<int8_t>()));
            case dtINT16:
                return res_tp(new fake_op::fake_setting_tp<int16_t>(obj.cast<int16_t>()));
            case dtINT32:
                return res_tp(new fake_op::fake_setting_tp<int32_t>(obj.cast<int32_t>()));
            case dtFLOAT32:
                return res_tp(new fake_op::fake_setting_tp<float>(obj.cast<float>()));
            case dtFLOAT64:
                return res_tp(new fake_op::fake_setting_tp<double>(obj.cast<double>()));
            case dtSTR:
                return res_tp(new fake_op::fake_setting_tp<string>(obj.cast<string>()));
            }
        }
        ostringstream ss;
        ss << "cannot parse setting type " << datatype_name_from_type(constraint->dtype) << "("
           << (constraint->ctp == ENUM_SETTING_CONSTRAINT_REPEATED ? "repeat" : "single") << ")";
        throw std::runtime_error(ss.str());
    }

    void proxy::register_python_module(py::module *module)
    {
        /* art.proxy */
        py::module &m = *module;
        py::class_<python_module_proxy>(m, "proxy")
            .def(py::init([](std::string name) {
                void *addr = nart::symbol::lookup_symbol(name + "_module_tp");
                if (nullptr == addr)
                    throw std::runtime_error("cannot find module '" + name + "'");
                return new python_module_proxy((const module_t *)addr);
            }))
            .def_property_readonly(
                "op_group", [](const python_module_proxy &self) { return self.op_group(); })
            .def_property_readonly(
                "op_tps", [](const python_module_proxy &self) { return self.op_tps(); })
            .def(
                "constraints_of",
                [](const python_module_proxy &self, std::string op_tp) {
                    return self.op_tp_constraints(op_tp);
                })
            .def_static("proxy_init", [](py::handle obj) {
                nart::fake_python_cls *true_obj = obj.cast<nart::fake_python_cls *>();
                auto proxy = obj.attr("module").attr("true_obj").cast<python_module_proxy *>();
                // auto dtype =
                // obj.attr("_inputs").cast<py::list>()[0].attr("dtype").cast<std::string>();
                std::string dtype = "float32";
                if (obj.attr("_inputs").cast<py::list>().size() > 0)
                    dtype
                        = obj.attr("_inputs").cast<py::list>()[0].attr("dtype").cast<std::string>();
                auto tp = proxy
                              ->search_op_tp_entry(
                                  obj.attr("__class__").attr("__name__").cast<std::string>())
                              ->tp;
                auto inferer = wrap_op_infer(tp, nart::fake_tensor::dtype_str2tp(dtype));
                true_obj->proxy_infer = inferer;
                true_obj->create_post_ = [obj, tp](fake_op *op) {
                    for (auto item : obj.attr("get_settings")()) {
                        uint32_t itm = item.cast<py::tuple>()[0].cast<std::uint32_t>();
                        for (const setting_constraint_t *p = tp->constraints;
                             p->item != SETTING_END; ++p) {
                            if (p->item == itm) {
                                auto set = parse_setting(p, item.cast<py::tuple>()[1]);
                                op->set_setting(itm, move(set));
                                break;
                            }
                        }
                    }
                };
            });
    }

}
}
